OldSneerJaw/couchster
# Issue 24: Regular expression pattern constraint for attachment filenames

# Description

When included in a document definition, the `filenameRegexPattern` constraint verifies that all attachments for that document type conform to the specified regular expression pattern.

# Testing

1. Generate a validation function from `attachment-constraints-doc-definitions.js`:

```bash
./make-validation-function -j test/resources/attachment-constraints-doc-definitions.js build/test-validation-function.txt
```

2. Copy and paste the validation function into a CouchDB design document.
3. Create a test document:

```
PUT /test/attachment-regex-test HTTP/1.1
Host: localhost:5984
Content-Type: application/json
Authorization: Basic ********

{
  "type": "staticAttachmentFilenameRegexPatternDoc"
}
```

4. Add an attachment with a valid filename to the document (it should succeed with 201 Created):

```
PUT /test/attachment-regex-test/foo.xls?rev=1-433aae3207c36b04bfd62a4a670d88a8 HTTP/1.1
Host: localhost:5984
Content-Type: text/plain
Authorization: Basic ********

This isn't actually an Excel document, despite the file extension, but it doesn't matter because the document type doesn't validate content type of attachments.
```

5. Attempt to add an attachment with an invalid filename (it should be rejected with 403 Forbidden):

```
PUT /test/attachment-regex-test/bar.invalid?rev=2-65733eae2b55e9ac6ce6a8830f843160 HTTP/1.1
Host: localhost:5984
Content-Type: text/plain
Authorization: Basic ********

The file name doesn't match the regex: /^(foo|bar)\.(xls|xlsx)$/
```

# Related Issue

* GitHub issue link: #24
